src/object_store/src/object/s3.rs

```diff
@@ -12,22 +12,259 @@
+    client: Arc<Client>,
+    bucket: String,
+    /// The key of the object.
+    key: String,
+    /// The identifier of multipart upload task for S3.
+    upload_id: String,
+    /// Next part ID.
+    next_part_id: PartId,
+    /// Join handles for part uploads.
+    join_handles: Vec<JoinHandle<ObjectResult<()>>>,
+    /// Parts that are already uploaded to S3.
+    uploaded_parts: Arc<Mutex<Vec<(PartId, UploadPartOutput)>>>,
+    /// Buffer for bytes.
+    buf: Vec<Bytes>,
+    /// Length of the data that have not been uploaded to S3.
+    not_uploaded_len: usize,
+    /// Length of the data that exceeds the current part to be uploaded in the buffer.
+    next_part_len: usize,
+    /// The data included in the next part are `buf[..part_end]`.
+    part_end: usize,
+    /// To record metrics for uploading part.
+    metrics: Arc<ObjectStoreMetrics>,
+}
+
+impl S3StreamingUploader {
+    pub fn new(
+        client: Arc<Client>,
+        bucket: String,
+        key: String,
+        upload_id: String,
+        metrics: Arc<ObjectStoreMetrics>,
+    ) -> S3StreamingUploader {
+        Self {
+            client,
+            bucket,
+            key,
+            upload_id,
+            next_part_id: MIN_PART_ID,
+            join_handles: Default::default(),
+            uploaded_parts: Default::default(),
+            buf: Default::default(),
+            not_uploaded_len: 0,
+            next_part_len: 0,
+            part_end: 0,
+            metrics,
+        }
+    }
+
+    fn upload_part(&mut self, data: Vec<Bytes>, len: usize) {
+        let part_id = self.next_part_id;
+        self.next_part_id += 1;
+        let client_cloned = self.client.clone();
+        let uploaded_parts_cloned = self.uploaded_parts.clone();
+        let bucket = self.bucket.clone();
+        let key = self.key.clone();
+        let upload_id = self.upload_id.clone();
+        let metrics = self.metrics.clone();
+        metrics
+            .operation_size
+            .with_label_values(&["s3_upload_part"])
+            .observe(len as f64);
+        self.join_handles.push(tokio::spawn(async move {
+            let timer = metrics
+                .operation_latency
+                .with_label_values(&["s3_upload_part"])
+                .start_timer();
+            let upload_output = client_cloned
+                .upload_part()
+                .bucket(bucket)
+                .key(key)
+                .upload_id(upload_id)
+                .part_number(part_id)
+                .body(get_upload_body(data))
+                .content_length(len as i64)
+                .send()
+                .await?;
+            timer.observe_duration();
+            uploaded_parts_cloned
+                .lock()
+                .map_err(ObjectError::internal)?
+                .push((part_id, upload_output));
+            Ok(())
+        }));
+    }
+
+    async fn flush_and_complete(&mut self) -> ObjectResult<()> {
+        self.upload_part(
```

Comment:
skip the upload if not_uploaded_len == 0

Reply:
Should we skip upload for `upload` as well?

Reply:
After a second thought, I realize that we will never have an empty part (not_uploaded_len == 0) when calling flush_and_complete because we keep one part + min_part_size part in the buf. Therefore, we should return an error if `not_uploaded_len  == 0`.

For normal `upload`, there is a difference between uploading an empty object (later get returns Ok(empty_object)) and not uploading the object (later get returns Err) at all. In our case, we will never upload an empty object so I think it is better to return an error rather than making it a no-op

Reply:
Cool